### ISSUE TITLE
Auto-continue comments

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#174]
It would be really nice if pressing Enter auto-continued comments. Examples (with `^` representing the cursor position when pressing Enter):

```
    // A comment^
```

to:

```
    // A comment
    // ^
```

and

```
    /* A comment^ */
```

to:

```
    /* A comment
     * ^ 
     */
```

or something similar.

It's not entirely clear whether this should be the job of the server or the client. I notice there is a ["textDocument/onTypeFormatting"](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#textDocument_onTypeFormatting) request, perhaps that would be a good fit?

**Comment by ilya-biryukov** [comment]
In fact, this is already implemented. In VSCode it's off by default, you'll need to set`"editor.formatOnType": false`.

**Comment by ilya-biryukov** [comment]
However, it does not do one of the suggested examples:
```
// Foo ^
```
Will get turned into 
```
// Foo
^
```

In a similar situation inside the comment, this would do what you'd expect:
```
// Foo^
// Bar
```
gets turned into
```
// Foo
// ^
// Bar
```

I would also vouch to extend existing comment when adding a newline at the end of it, but I believe @sam-mccall had a different opinion there.

**Comment by sam-mccall** [comment]
Right, break within a line comment should work. For now at least breaking at the end of a line comment doesn't continue the comment (deliberately). I don't think we implemented box/block comments (drawing aligned stars), but that could be added.

I think for that case (at the end of a line) there are examples that cut either way. e.g.

```
void send() {
  request_opts.retry = false; // retry on the caller side^
}
```

Editors tend to make a choice about this, and when they do, we respect it. (The editor can insert // and then send onTypeFormatting and clangd will do the right thing).

We could attempt some heuristics (capital letter without punctuation) but I'm not sure whether the utility would outweigh the lack of predictability and exceptions.

**Comment by HighCommander4** [comment]
> I think for that case (at the end of a line) there are examples that cut either way. e.g.
> 
> ```
> void send() {
>   request_opts.retry = false; // retry on the caller side^
> }
> ```

Thanks, this is a good example.

Editors do indeed have a mix of behaviours here; for example, Eclipse does not extend line comments, and vim does only if the comment is the only thing on the line (so it wouldn't extend in the above example). I'm happy to defer to the editor here.

Perhaps we can make this issue specific to block comments then.

**Comment by travisdowns** [comment]
> In fact, this is already implemented. In VSCode it's off by default, you'll need to set`"editor.formatOnType": false`.

Are you suggesting that if this is enabled, it should work when extending a block commend when `clangd` is enabled as the language server?

It doesn't work for me. That is, putting the cursor at the end of `here`:

```c++
/**
 * here
 */
```

and pressing enter results in:

```c++
/**
 * here
 
 */
```

Rather than

```c++
/**
 * here
 *  
 */
```

 as I would expect if it were working.

**Comment by HighCommander4** [comment]
@travisdowns It looks like the "this is already implemented" part only applies to `//` style comments  (and even then there is a limitation that an <enter> at the end of the last line of a multi-line `//` comment does not extend the comment -- it only works in the middle of a line, or at the end of a line that's not the last).

**Comment by travisdowns** [comment]
Though it was not the answer I was hoping for, I appreciate the clarification, thank you!

**Comment by starball5** [comment]
Related cpptools feature-request that got implemented in its 0.13.0: https://github.com/Microsoft/vscode-cpptools/issues/579

Related VS Code FR in October 2023 Milestone: https://github.com/microsoft/vscode/issues/180551

**Comment by LeventCelik** [comment]
I'd like to revive this. Very minor inconvenience, but too often encountered to disregard.

**Comment by HighCommander4** [comment]
> I'd like to revive this. Very minor inconvenience, but too often encountered to disregard.

Are you interested in working on a patch?

**Comment by LeventCelik** [comment]
Sure. Is one already being worked on, or shall I start myself?

**Comment by HighCommander4** [comment]
> Sure.

Thanks!

> Is one already being worked on

Not that I'm aware of.

Let me know if you need some pointers to where in the code to start looking.

**Comment by LeventCelik** [comment]
> Let me know if you need some pointers to where in the code to start looking.

Thanks! I'm not familiar with the codebase, so any pointers would be great. I have joined the Discord channel, we can continue there if that's more convenient.

**Comment by JVApen** [comment]
> ```
>     /* A comment^ */
> ```
> 
> to:
> 
> ```
>     /* A comment
>      * ^ 
>      */
> ```

Our style guide disallows block comments with leading `*`.
```
    /*
    A comment
    ^ 
    */
```
Would be inline with our style guide.

Not saying this should be supported, though it might be useful to have an option to disable the extra `*` such that you can get:
```
    /* A comment
    ^ 
    */
```

**Comment by JVApen** [comment]
It might also be worth mentioning that `//` is not the only comment marker of relevance. `///`, `//!<` and a few others should also be respected. You basically want the same behavior as when clang-format splits a comment line in 2. (See https://github.com/llvm/llvm-project/blob/f2e244f443c99f814216bef37a82f1b8d3e5975d/clang/lib/Format/BreakableToken.cpp#L32)
